opencart 1.5.6.4/upload/catalog/controller/smailyforopencart/cron_customers.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+/**
+ * Samaly subscribers synchronization.
+ */
+class ControllerSmailyForOpencartCronCustomers extends Controller {
+
+    public function index() {
+        // Customer model.
+        $this->load->model('account/customer');
+        // Load Smaily settings.
+        $this->load->model('setting/setting');
+        // Load Smaily helper.
+        $this->load->model('smailyforopencart/helper');
+        
+        $settings = $this->model_setting_setting->getSetting('smaily')['smaily_customer_sync'];
+
+        // Validate cron token.
+        if (!array_key_exists('token', $settings) ||
+            empty($this->request->get['token']) ||
+            $settings['token'] !== $this->request->get['token']
+        ) {
+            die('Unauthorized');
+        }
+        if ($settings['enabled'] == 1) {
+
+            // Get unsubscribers from smaily.
+            $unsubscribers_filter = array(
+                'list' => 2,
+            );
+
+            $unsubscribers = $this->model_smailyforopencart_helper->apiCall('contact', $unsubscribers_filter);
+            // Collect unsubscriber emails.
+            $unsubscribers_emails = [];
+            foreach ($unsubscribers as $unsubscriber) {
+                array_push($unsubscribers_emails, $unsubscriber['email']);
+            }
+            // Remove subscribed status from unsubscribed customers.
+            $subscribers = $this->model_smailyforopencart_helper->getSubscribedCustomers();
+            foreach ($subscribers as $subscriber) {
+                if (in_array($subscriber['email'], $unsubscribers_emails)) {
+                    $this->model_smailyforopencart_helper->unsubscribeCustomer($subscriber['customer_id']);
+                }
+            }
+            
+            // Get all users with subscribed status.
+            $subscribers = $this->model_smailyforopencart_helper->getSubscribedCustomers();
+            if (empty($subscribers)) {
+                $this->log->write('Smaily subscriber sync without customers.');
+                echo 'No subscribers to sync.';
+            } else {
+                // Prepare list for smaily.
+                $list = [];
+                foreach ($subscribers as $subscriber) {
+                    // Get customer info based of selected fields from admin.
+                    $sync_fields = $this->model_smailyforopencart_helper->getSyncFields();
+                    $customer = [];
+                    foreach ($sync_fields as $field) {
+                        $customer[$field] = $subscriber[$field];
```

Comment:
Can there be a case where `$subscriber[$field]` does not exist?

Reply:
no, getSyncFields() returns email as default if there are no fields selected